llap-common/src/java/org/apache/hadoop/hive/llap/AsyncResponseHandler.java

```diff
@@ -0,0 +1,126 @@
+/*
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package org.apache.hadoop.hive.llap;
+
+import com.google.protobuf.Message;
+import org.apache.hadoop.util.concurrent.AsyncGet;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import java.util.ArrayList;
+import java.util.Iterator;
+import java.util.List;
+import java.util.Queue;
+import java.util.concurrent.ConcurrentLinkedDeque;
+import java.util.concurrent.ExecutorService;
+import java.util.concurrent.Executors;
+import java.util.concurrent.TimeUnit;
+import java.util.concurrent.atomic.AtomicBoolean;
+
+public class AsyncResponseHandler {
+
+  private static final Logger LOG = LoggerFactory.getLogger(AsyncResponseHandler.class);
+
+  private final AtomicBoolean isShutdown = new AtomicBoolean(false);
+
+  private final AsyncPbRpcProxy.RequestManager requestManager;
+  private final ExecutorService responseWaitingService = Executors.newSingleThreadExecutor();
+  private final ExecutorService responseExecutorService = Executors.newSingleThreadExecutor();
+  private final Queue<AsyncPbRpcProxy.AsyncCallableRequest<Message, Message>> asyncResponseFutureQueue = new ConcurrentLinkedDeque<>();
+
+  public AsyncResponseHandler(AsyncPbRpcProxy.RequestManager requestManager) {
+    this.requestManager = requestManager;
+  }
+
+  public void start() {
+    responseWaitingService.submit(new AsyncResponseHandlerRunnable());
+  }
+
+  public void addToAsyncResponseFutureQueue(AsyncPbRpcProxy.AsyncCallableRequest<Message, Message> request) {
+    asyncResponseFutureQueue.add(request);
+  }
+
+  public void shutdown() {
+    isShutdown.set(true);
+    responseExecutorService.shutdown();
+    responseWaitingService.shutdown();
+  }
+
+  private final class AsyncResponseHandlerRunnable implements Runnable {
+
+    private final List<AsyncPbRpcProxy.AsyncCallableRequest<Message, Message>> asyncResponseFutures = new ArrayList<>();
+
+    @Override
+    public void run() {
+      while (!isShutdown.get()) {
```

Comment:
Isn't it a tight while loop? As in, it will consume 1 CPU @ 100% all the time if empty. Can you plz fix that?